File: version_java/src/main/java/com/example/version_java/data/repo/CriminalRepositoryImpl.java
package com.example.version_java.data.repo;

import androidx.annotation.NonNull;

import com.example.version_java.data.source.local.CrimianlLocalDataSource;
import com.example.version_java.data.source.remote.CriminalRemoteDataSource;
import com.example.version_java.room.entity.CriminalEntity;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import kotlin.jvm.functions.Function1;

public class CriminalRepositoryImpl implements CriminalRepository {

    private final CriminalRemoteDataSource criminalRemoteDataSource;

    private final CrimianlLocalDataSource crimianlLocalDataSource;

    @Override
    public void getRemoteCriminals(@NonNull Function1 onSuccess, @NonNull Function1 onFailure) {
        criminalRemoteDataSource.getRemoteCriminals(onSuccess, onFailure);
    }

    @Override
    public void getLocalCriminals(@NonNull Function1 onSuccess, @NonNull Function1 onFailure) {
        crimianlLocalDataSource.getLocalCriminals(onSuccess, onFailure);
    }


    @Override
    public void registerCriminalEntity(ArrayList<CriminalEntity> entityList, @NonNull Function1 result) {
        crimianlLocalDataSource.registerCriminalEntity(entityList, result);
    }

    @Override
    public void getCriminalEntity(@NonNull String name, @NonNull Function1 onSuccess, @NonNull Function1 onFailure) {
        crimianlLocalDataSource.getCriminalEntity(name, onSuccess, onFailure);
    }

    @Inject
    CriminalRepositoryImpl(
            CriminalRemoteDataSource criminalRemoteDataSource,
            CrimianlLocalDataSource crimianlLocalDataSource
    ) {
        this.criminalRemoteDataSource = criminalRemoteDataSource;
        this.crimianlLocalDataSource = crimianlLocalDataSource;
    }
}

File: app/src/main/java/com/example/crimenotification/data/repo/CriminalRepositoryImpl.kt
package com.example.crimenotification.data.repo

import com.example.crimenotification.data.source.local.CriminalLocalDataSource
import com.example.crimenotification.data.source.remote.CriminalRemoteDataSource
import com.example.crimenotification.network.response.CriminalResponse
import com.example.crimenotification.room.entity.CriminalEntity
import com.example.crimenotification.util.Result
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import javax.inject.Inject

class CriminalRepositoryImpl @Inject constructor(
    private val criminalRemoteDataSource: CriminalRemoteDataSource,
    private val criminalLocalDataSource: CriminalLocalDataSource
) : CriminalRepository {

    override suspend fun getRemoteCriminals(): Result<List<CriminalResponse>> =
        withContext(Dispatchers.IO) {
            return@withContext criminalRemoteDataSource.getCriminals()
        }

    override suspend fun getLocalCriminals(): Result<List<CriminalEntity>> =
        withContext(Dispatchers.IO) {
            return@withContext criminalLocalDataSource.getLocalCriminals()
        }


    override suspend fun registerCriminalEntity(entityList: List<CriminalEntity>): Boolean =
        withContext(Dispatchers.IO) {
            return@withContext criminalLocalDataSource.registerCriminalEntity(entityList)
        }

    override suspend fun getCriminalEntity(name: String): Result<CriminalEntity> =
        withContext(Dispatchers.IO) {
            return@withContext criminalLocalDataSource.getCriminalEntity(name)
        }
}